File: db-4.8.30/examples_java/src/db/BtRecExample.java
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997-2009 Oracle.  All rights reserved.
 *
 * $Id$
 */


package db;

import com.sleepycat.db.*;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.io.PrintStream;

public class BtRecExample {
    static final String progname =  "BtRecExample"; // Program name.
    static final String database =  "access.db";
    static final String wordlist =  "../test/wordlist";

    BtRecExample(BufferedReader reader)
        throws DatabaseException, IOException, FileNotFoundException {

        OperationStatus status;

        // Remove the previous database.
        File f = new File(database);
        f.delete();

        DatabaseConfig config = new DatabaseConfig();

        config.setErrorStream(System.err);
        config.setErrorPrefix(progname);
        config.setPageSize(1024);           // 1K page sizes.

        config.setBtreeRecordNumbers(true);
        config.setType(DatabaseType.BTREE);
        config.setAllowCreate(true);
        db = new Database(database, null, config);

        //
        // Insert records into the database, where the key is the word
        // preceded by its record number, and the data is the same, but
        // in reverse order.
        //

        for (int cnt = 1; cnt <= 1000; ++cnt) {
            String numstr = String.valueOf(cnt);
            while (numstr.length() < 4)
                numstr = "0" + numstr;
            String buf = numstr + '_' + reader.readLine();
            StringBuffer rbuf = new StringBuffer(buf).reverse();

            StringEntry key = new StringEntry(buf);
            StringEntry data = new StringEntry(rbuf.toString());

            status = db.putNoOverwrite(null, key, data);
            if (status != OperationStatus.SUCCESS &&
                status!= OperationStatus.KEYEXIST)
                throw new DatabaseException("Database.put failed " + status);
        }
    }

    void run() throws DatabaseException {
        int recno;
        OperationStatus status;

        // Acquire a cursor for the database.
        cursor = db.openCursor(null, null);

        //
        // Prompt the user for a record number, then retrieve and display
        // that record.
        //
        InputStreamReader reader = new InputStreamReader(System.in);

        for (;;) {
            // Get a record number.
            String line = askForLine(reader, System.out, "recno #> ");
            if (line == null)
                break;

            try {
                recno = Integer.parseInt(line);
            } catch (NumberFormatException nfe) {
                System.err.println("Bad record number: " + nfe);
                continue;
            }

            //
            // Start with a fresh key each time, the db.get() routine returns
            // the key and data pair, not just the key!
            //
            RecnoStringEntry key = new RecnoStringEntry(recno, 4);
            RecnoStringEntry data = new RecnoStringEntry(4);

            status = cursor.getSearchRecordNumber(key, data, null);
            if (status != OperationStatus.SUCCESS)
                throw new DatabaseException("Cursor.setRecno failed: " + status);

            // Display the key and data.
            show("k/d\t", key, data);

            // Move the cursor a record forward.
            status = cursor.getNext(key, data, null);
            if (status != OperationStatus.SUCCESS)
                throw new DatabaseException("Cursor.getNext failed: " + status);

            // Display the key and data.
            show("next\t", key, data);

            RecnoStringEntry datano = new RecnoStringEntry(4);

            //
            // Retrieve the record number for the following record into
            // local memory.
            //
            status = cursor.getRecordNumber(datano, null);
            if (status != OperationStatus.SUCCESS &&
                status != OperationStatus.NOTFOUND &&
                status != OperationStatus.KEYEMPTY)
                throw new DatabaseException("Cursor.get failed: " + status);
            else {
                recno = datano.getRecordNumber();
                System.out.println("retrieved recno: " + recno);
            }
        }

        cursor.close();
        cursor = null;
    }

    //
    // Print out the number of records in the database.
    //
    void stats() throws DatabaseException {
        BtreeStats stats;

        stats = (BtreeStats)db.getStats(null, null);
        System.out.println(progname + ": database contains " +
               stats.getNumData() + " records");
    }

    void show(String msg, RecnoStringEntry key, RecnoStringEntry data)
        throws DatabaseException {

        System.out.println(msg + key.getString() + ": " + data.getString());
    }

    public void shutdown() throws DatabaseException {
        if (cursor != null) {
            cursor.close();
            cursor = null;
        }
        if (db != null) {
            db.close();
            db = null;
        }
    }

    public static void main(String[] argv) {
        try {
            // Open the word database.
            FileReader freader = new FileReader(wordlist);

            BtRecExample app = new BtRecExample(new BufferedReader(freader));

            // Close the word database.
            freader.close();
            freader = null;

            app.stats();
            app.run();
        } catch (FileNotFoundException fnfe) {
            System.err.println(progname + ": unexpected open error " + fnfe);
            System.exit (1);
        } catch (IOException ioe) {
            System.err.println(progname + ": open " + wordlist + ": " + ioe);
            System.exit (1);
        } catch (DatabaseException dbe) {
            System.err.println("Exception: " + dbe);
            System.exit(dbe.getErrno());
        }

        System.exit(0);
    }

    // Prompts for a line, and keeps prompting until a non blank
    // line is returned.  Returns null on error.
    //
    public static String askForLine(InputStreamReader reader,
                                    PrintStream out, String prompt) {
        String result = "";
        while (result != null && result.length() == 0) {
            out.print(prompt);
            out.flush();
            result = getLine(reader);
        }
        return result;
    }

    // Not terribly efficient, but does the job.
    // Works for reading a line from stdin or a file.
    // Returns null on EOF.  If EOF appears in the middle
    // of a line, returns that line, then null on next call.
    //
    public static String getLine(InputStreamReader reader) {
        StringBuffer b = new StringBuffer();
        int c;
        try {
            while ((c = reader.read()) != -1 && c != '\n') {
                if (c != '\r')
                    b.append((char)c);
            }
        } catch (IOException ioe) {
            c = -1;
        }

        if (c == -1 && b.length() == 0)
            return null;
        else
            return b.toString();
    }

    private Cursor cursor;
    private Database db;

    // Here's an example of how you can extend DatabaseEntry in a
    // straightforward way to allow easy storage/retrieval of strings.
    // We've declared it as a static inner class, but it need not be.
    //
    static class StringEntry extends DatabaseEntry {
        StringEntry() {}

        StringEntry(String value) {
            setString(value);
        }

        void setString(String value) {
            byte[] data = value.getBytes();
            setData(data);
            setSize(data.length);
        }

        String getString() {
            return new String(getData(), 0, getSize());
        }
    }

    // Here's an example of how you can extend DatabaseEntry to store
    // (potentially) both recno's and strings in the same structure.
    //
    static class RecnoStringEntry extends DatabaseEntry {
        RecnoStringEntry(int maxsize) {
            this(0, maxsize);     // let other constructor do most of the work
        }

        RecnoStringEntry(int value, int maxsize) {
            arr = new byte[maxsize];
            setRecordNumber(value);
            setSize(arr.length);
        }

        RecnoStringEntry(String value) {
            byte[] data = value.getBytes();
            setData(data);                // use our local array for data
            setUserBuffer(data.length, true);
        }

        void setString(String value) {
            byte[] data = value.getBytes();
            setData(data);
            setSize(data.length);
        }

        String getString() {
            return new String(getData(), getOffset(), getSize());
        }

        byte[] arr;
    }
}

File: db-4.8.30/examples_cxx/BtRecExample.cpp
/*-
 * See the file LICENSE for redistribution information.
 *
 * Copyright (c) 1997-2009 Oracle.  All rights reserved.
 *
 * $Id$
 */

#include <sys/types.h>

#include <errno.h>
#include <iostream>
#include <iomanip>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <db_cxx.h>

using std::cout;
using std::cerr;

#define DATABASE    "access.db"
#define WORDLIST    "../test/wordlist"

const char *progname = "BtRecExample";      // Program name.

class BtRecExample
{
public:
    BtRecExample(FILE *fp);
    ~BtRecExample();
    void run();
    void stats();
    void show(const char *msg, Dbt *key, Dbt *data);

private:
    Db *dbp;
    Dbc *dbcp;
};

BtRecExample::BtRecExample(FILE *fp)
{
    char *p, *t, buf[1024], rbuf[1024];
    int ret;

    // Remove the previous database.
    (void)remove(DATABASE);

    dbp = new Db(NULL, 0);

    dbp->set_error_stream(&cerr);
    dbp->set_errpfx(progname);
    dbp->set_pagesize(1024);            // 1K page sizes.

    dbp->set_flags(DB_RECNUM);          // Record numbers.
    dbp->open(NULL, DATABASE, NULL, DB_BTREE, DB_CREATE, 0664);

    //
    // Insert records into the database, where the key is the word
    // preceded by its record number, and the data is the same, but
    // in reverse order.
    //

    for (int cnt = 1; cnt <= 1000; ++cnt) {
        (void)sprintf(buf, "%04d_", cnt);
        if (fgets(buf + 4, sizeof(buf) - 4, fp) == NULL)
            break;
        u_int32_t len = (u_int32_t)strlen(buf);
        buf[len - 1] = '\0';
        for (t = rbuf, p = buf + (len - 2); p >= buf;)
            *t++ = *p--;
        *t++ = '\0';

        // As a convenience for printing, we include the null terminator
        // in the stored data.
        //
        Dbt key(buf, len);
        Dbt data(rbuf, len);

        if ((ret = dbp->put(NULL, &key, &data, DB_NOOVERWRITE)) != 0) {
            dbp->err(ret, "Db::put");
            if (ret != DB_KEYEXIST)
                throw DbException(ret);
        }
    }
}

BtRecExample::~BtRecExample()
{
    if (dbcp != 0)
        dbcp->close();
    dbp->close(0);
    delete dbp;
}

//
// Print out the number of records in the database.
//
void BtRecExample::stats()
{
    DB_BTREE_STAT *statp;

    dbp->stat(NULL, &statp, 0);
    cout << progname << ": database contains "
         << (u_long)statp->bt_ndata << " records\n";

    // Note: must use free, not delete.
    // This struct is allocated by C.
    //
    free(statp);
}

void BtRecExample::run()
{
    db_recno_t recno;
    int ret;
    char buf[1024];

    // Acquire a cursor for the database.
    dbp->cursor(NULL, &dbcp, 0);

    //
    // Prompt the user for a record number, then retrieve and display
    // that record.
    //
    for (;;) {
        // Get a record number.
        cout << "recno #> ";
        cout.flush();
        if (fgets(buf, sizeof(buf), stdin) == NULL)
            break;
        recno = atoi(buf);

        //
        // Start with a fresh key each time,
        // the dbp->get() routine returns
        // the key and data pair, not just the key!
        //
        Dbt key(&recno, sizeof(recno));
        Dbt data;

        if ((ret = dbcp->get(&key, &data, DB_SET_RECNO)) != 0) {
            dbp->err(ret, "DBcursor->get");
            throw DbException(ret);
        }

        // Display the key and data.
        show("k/d\t", &key, &data);

        // Move the cursor a record forward.
        if ((ret = dbcp->get(&key, &data, DB_NEXT)) != 0) {
            dbp->err(ret, "DBcursor->get");
            throw DbException(ret);
        }

        // Display the key and data.
        show("next\t", &key, &data);

        //
        // Retrieve the record number for the following record into
        // local memory.
        //
        data.set_data(&recno);
        data.set_size(sizeof(recno));
        data.set_ulen(sizeof(recno));
        data.set_flags(data.get_flags() | DB_DBT_USERMEM);

        if ((ret = dbcp->get(&key, &data, DB_GET_RECNO)) != 0) {
            if (ret != DB_NOTFOUND && ret != DB_KEYEMPTY) {
                dbp->err(ret, "DBcursor->get");
                throw DbException(ret);
            }
        }
        else {
            cout << "retrieved recno: " << (u_long)recno << "\n";
        }
    }

    dbcp->close();
    dbcp = NULL;
}

//
// show --
//  Display a key/data pair.
//
void BtRecExample::show(const char *msg, Dbt *key, Dbt *data)
{
    cout << msg << (char *)key->get_data()
         << " : " << (char *)data->get_data() << "\n";
}

int
main()
{
    FILE *fp;

    // Open the word database.
    if ((fp = fopen(WORDLIST, "r")) == NULL) {
        fprintf(stderr, "%s: open %s: %s\n",
            progname, WORDLIST, db_strerror(errno));
        return (EXIT_FAILURE);
    }

    try {
        BtRecExample app(fp);

        // Close the word database.
        (void)fclose(fp);
        fp = NULL;

        app.stats();
        app.run();
    }
    catch (DbException &dbe) {
        cerr << "Exception: " << dbe.what() << "\n";
        return (EXIT_FAILURE);
    }

    return (EXIT_SUCCESS);
}
